### PR TITLE
Feature/local mempool test

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/invariants/mempool/MempoolSubmitAndCommitInvariant.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * Contributes to steady state by submitting client atoms every few seconds and
+ * validates that they have all been committed by all nodes.
+ */
+public class MempoolSubmitAndCommitInvariant implements TestInvariant {
+
+	private Maybe<TestInvariantError> submitAndCheckForCommit(Mempool mempool, Set<Observable<CommittedCommand>> allLedgers) {
+		ClientAtom clientAtom = mock(ClientAtom.class);
+
+		List<Maybe<TestInvariantError>> errors = allLedgers.stream()
+			.map(cmds ->
```
Comment (clstrfsck) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/invariants/mempool/MempoolSubmitAndCommitInvariant.java]: Wouldn't this be more predictably formatted like the below?
```
    .map(cmds -> cmds
        .filter(...)
        .timeout(...
        .etc(....)
```

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/invariants/mempool/MempoolSubmitAndCommitInvariant.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+					.doOnComplete(() -> System.out.println("Committed " + clientAtom))
+					.onErrorReturn(e -> new TestInvariantError(e.getMessage() + " " + clientAtom))
+			)
+			.collect(Collectors.toList());
+
+		return Maybe.merge(errors).firstElement().doOnSubscribe(d -> {
+			try {
+				mempool.addAtom(clientAtom);
+			} catch (MempoolDuplicateException | MempoolFullException e) {
+				e.printStackTrace();
```
Comment (clstrfsck) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/invariants/mempool/MempoolSubmitAndCommitInvariant.java:63]: This really needs fixing.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/tests/consensus_executor_localmempool/MempoolSanityTest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		assertThat(results).hasEntrySatisfying("mempool", error -> assertThat(error).isPresent());
+	}
+
+	@Test
+	public void when_submitting_items_to_mempool__then_they_should_get_executed() {
+		SimulationTest simulationTest = bftTestBuilder
+			.executorAndMempool()
+			.build();
+		Map<String, Optional<TestInvariantError>> results = simulationTest.run(1, TimeUnit.MINUTES);
+		assertThat(results).allSatisfy((name, err) -> AssertionsForClassTypes.assertThat(err).isEmpty());
```
Comment (clstrfsck) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/tests/consensus_executor_localmempool/MempoolSanityTest.java:58]: Not obvious to me how this checks that things actually get executed?

Not clear how absence of errors is the same thing as presence of non-null commits here.

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexMetadata.java [hunk 4]
```diff
@@ -129,6 +161,32 @@ public Hash getId() {
...
 	}
 
+	@JsonProperty("validator_set")
+	@DsonOutput(Output.ALL)
+	private Map<String, UInt256> getValidatorSetJson() {
+		if (validatorSet == null) {
+			return null;
+		}
+		return validatorSet.getValidators().stream()
+			.collect(ImmutableMap.toImmutableMap(v -> v.getNode().getKey().toBase64(), BFTValidator::getPower));
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/consensus/VertexMetadata.java]: We have pretty much exclusively used hex for pubkeys everywhere else.

If we want to change this, also needs changing in ECDSASignatures, TimestampedECDSASignatures and GenerateUniverses.

File: radixdlt/src/main/java/com/radixdlt/syncer/CommittedCommands.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import com.radixdlt.middleware2.CommittedAtom;
+import com.radixdlt.syncer.SyncExecutor.CommittedCommand;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Temporary class to hold committed results
+ * TODO: Remove class
+ */
+public class CommittedCommands {
```
Comment (clstrfsck) [inline radixdlt/src/main/java/com/radixdlt/syncer/CommittedCommands.java]: Should be final.

File: radixdlt/src/test/java/com/radixdlt/ExecutionModuleTest.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+import com.radixdlt.syncer.SyncExecutor.CommittedStateSyncSender;
+import com.radixdlt.syncer.SyncExecutor.StateComputer;
+import com.radixdlt.syncer.SyncExecutor.SyncService;
+import java.util.Optional;
+import org.junit.Test;
+
+public class ExecutionModuleTest {
+	private static class ExternalExecutionModule extends AbstractModule {
+		@Override
+		@SuppressWarnings("unchecked")
```
Comment (clstrfsck) [inline radixdlt/src/test/java/com/radixdlt/ExecutionModuleTest.java]: This doesn't seem to be necessary?

File: radixdlt/src/main/java/org/radix/api/services/AtomsService.java [hunk 7]
```diff
@@ -107,10 +109,20 @@ public AtomsService(
...
+			return;
+		}
+
+		executedCommand
+			.ifSuccess(result -> this.processStoredEvent(executedCommand.getCommand(), result))
+			.ifError(e -> this.processException(executedCommand.getCommand(), e));
+	}
+
+	private void processStoredEvent(CommittedAtom committedAtom, Object result) {
+		ImmutableSet<EUID> indicies = (ImmutableSet<EUID>) result;
```
Comment (clstrfsck) [inline radixdlt/src/main/java/org/radix/api/services/AtomsService.java:124]: This is pretty horrible, and really should have a FIXME.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/network/SimulationNodes.java [hunk 8]
```diff
@@ -96,6 +105,11 @@ public RunningNetwork start() {
 		}
 
 		return new RunningNetwork() {
+			public EpochChange initialEpoch() {
```
Comment (clstrfsck) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/network/SimulationNodes.java:109]: Missing Override annotation.

File: radixdlt/src/test/java/com/radixdlt/StateComputerModuleTest.java [hunk 9]
```diff
@@ -49,14 +48,13 @@ protected void configure() {
 			Universe universe = mock(Universe.class);
```
Comment (clstrfsck) [inline radixdlt/src/test/java/com/radixdlt/StateComputerModuleTest.java:47]: SupressWarnings("unchecked") annotation here seems unnecessary.

**Review by clstrfsck — CHANGES_REQUESTED** [review]
Biggest issue here seems to be how we will serialize pubkeys when used as map keys and for display.  All the existing code uses hex, but this PR uses base64.  We need some changes in order to be consistent here, I don't have a strong preference which way we go, but it needs to be consistent.

Some additional minor things needing attention also.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/invariants/mempool/MempoolSubmitAndCommitInvariant.java (hunk 2 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/invariants/mempool/MempoolSubmitAndCommitInvariant.java:63]: Added a TODO comment

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/tests/consensus_executor_localmempool/MempoolSanityTest.java (hunk 3 above)
Comment (talekhinezh) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/tests/consensus_executor_localmempool/MempoolSanityTest.java:58]: The invariant checks that the submitted atom was committed (not just that it's non-null). Is that what you mean?

File: radixdlt/src/main/java/com/radixdlt/consensus/VertexMetadata.java (hunk 4 above)
Comment (talekhinezh) [reply radixdlt/src/main/java/com/radixdlt/consensus/VertexMetadata.java]: fair enough

**Pull request merged**